python/ray/data/_internal/arrow_ops/transform_pyarrow.py

```diff
@@ -45,6 +50,49 @@ def take_table(
     return table
 
 
+def unify_schemas(
+    schemas: List["pyarrow.Schema"],
+) -> "pyarrow.Schema":
+    """Version of `pyarrow.unify_schemas()` which also handles checks for
+    variable-shaped tensors in the given schemas."""
+    schemas_to_unify = []
+    schema_tensor_field_overrides = {}
+
+    if any(isinstance(type_, pyarrow.ExtensionType) for type_ in schemas[0].types):
+        # If we have pyarrow extension types that may potentially be variable shaped,
+        # examine the first schema to gather the columns that need type conversions.
+        for col_idx in range(len(schemas[0].types)):
+            tensor_array_types = []
+            for s in schemas:
+                col_type = s.types[col_idx]
+                if isinstance(col_type, pyarrow.ExtensionType):
+                    tensor_array_types.append(col_type)
```

Comment:
Supernit: Could condense into a list comprehension without sacrificing readability.
```suggestion
            tensor_array_types = [
                s.types[col_idx] for s in schemas if isinstance(s.types[col_idx], pyarrow.ExtensionType)
            ]
```